File: apps/tabtin-ios/Tabtin/Core/Workspace/ProjectModels.swift
import Foundation

/// Project 是团队协作场景；后端物理实现仍是 Space(type=team_space)。
/// 移动端只消费协作字段，不把本地目录或设备误建模为 Project 自身属性。
struct Project: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let organizationId: String
    let type: String?
    let name: String
    let description: String?
    let avatar: String?
    let color: String?
    let status: String?
    let isArchived: Bool?
    let visibility: String?
    let memberCount: Int?
    let primaryAgentId: String?
    let canManage: Bool?
    let lastActivityAt: String?
    let createdAt: String?
    let updatedAt: String?
    let myWorkspace: ProjectCompanionWorkspace?

    enum CodingKeys: String, CodingKey {
        case id, type, name, description, avatar, color, status, visibility
        case organizationId = "organization_id"
        case isArchived = "is_archived"
        case memberCount = "member_count"
        case primaryAgentId = "primary_agent_id"
        case canManage = "can_manage"
        case lastActivityAt = "last_activity_at"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case myWorkspace = "my_workspace"
    }

    var displayDescription: String? {
        guard let value = description?.trimmingCharacters(in: .whitespacesAndNewlines), !value.isEmpty else {
            return nil
        }
        return value
    }

    var displayTime: String? {
        guard let value = lastActivityAt ?? updatedAt ?? createdAt else { return nil }
        return RelativeTime.format(value)
    }
}

struct ProjectCompanionWorkspace: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let name: String?
    let agentId: String?
    let executionAgentId: String?
    let workingDir: String?
    let controlDeviceId: String?
    let controlDeviceStatus: String?
    let isCompanion: Bool?

    enum CodingKeys: String, CodingKey {
        case id, name
        case agentId = "agent_id"
        case executionAgentId = "execution_agent_id"
        case workingDir = "working_dir"
        case controlDeviceId = "control_device_id"
        case controlDeviceStatus = "control_device_status"
        case isCompanion = "is_companion"
    }
}

struct ProjectListResponse: Decodable, Sendable {
    let projects: [Project]
    let total: Int?
}

struct ProjectPrimaryAgentResponse: Decodable, Sendable {
    let projectId: String
    let primaryAgentId: String?

    enum CodingKeys: String, CodingKey {
        case projectId = "project_id"
        case primaryAgentId = "primary_agent_id"
    }
}

struct PendingProjectInvitation: Codable, Identifiable, Hashable, Sendable {
    let projectId: String
    let projectName: String
    let organizationId: String
    let role: String
    let inviterName: String
    let invitedAt: String?

    var id: String { projectId }

    enum CodingKeys: String, CodingKey {
        case role
        case projectId = "project_id"
        case projectName = "project_name"
        case organizationId = "organization_id"
        case inviterName = "inviter_name"
        case invitedAt = "invited_at"
    }
}

struct PendingProjectInvitationListResponse: Decodable, Sendable {
    let invitations: [PendingProjectInvitation]
    let total: Int?
}

struct ProjectDiscussion: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let organizationId: String
    let spaceId: String?
    let spaceName: String?
    let isTeamSpaceChannel: Bool
    let name: String
    let memberCount: Int?
    let isArchived: Bool
    let lastMessageAt: String?
    let lastMessagePreview: String?
    let unreadCount: Int

    enum CodingKeys: String, CodingKey {
        case id, name
        case organizationId = "organization_id"
        case spaceId = "space_id"
        case spaceName = "space_name"
        case isTeamSpaceChannel = "is_team_space_channel"
        case memberCount = "member_count"
        case isArchived = "is_archived"
        case lastMessageAt = "last_message_at"
        case lastMessagePreview = "last_message_preview"
        case unreadCount = "unread_count"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id)
        organizationId = try c.decode(String.self, forKey: .organizationId)
        spaceId = try c.decodeIfPresent(String.self, forKey: .spaceId)
        spaceName = try c.decodeIfPresent(String.self, forKey: .spaceName)
        isTeamSpaceChannel = try c.decodeIfPresent(Bool.self, forKey: .isTeamSpaceChannel) ?? false
        name = try c.decodeIfPresent(String.self, forKey: .name) ?? ""
        memberCount = try c.decodeIfPresent(Int.self, forKey: .memberCount)
        isArchived = try c.decodeIfPresent(Bool.self, forKey: .isArchived) ?? false
        lastMessageAt = try c.decodeIfPresent(String.self, forKey: .lastMessageAt)
        lastMessagePreview = try c.decodeIfPresent(String.self, forKey: .lastMessagePreview)
        unreadCount = try c.decodeIfPresent(Int.self, forKey: .unreadCount) ?? 0
    }
}

struct ProjectActivityEvent: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let eventType: String
    let actorUserId: String?
    let actorName: String?
    let targetType: String?
    let targetId: String?
    let targetName: String?
    let metadata: [String: AnyCodable]?
    let createdAt: String

    enum CodingKeys: String, CodingKey {
        case id, metadata
        case eventType = "event_type"
        case actorUserId = "actor_user_id"
        case actorName = "actor_name"
        case targetType = "target_type"
        case targetId = "target_id"
        case targetName = "target_name"
        case createdAt = "created_at"
    }

    var displayTime: String? { RelativeTime.format(createdAt) }
}

struct ProjectActivityListResponse: Decodable, Sendable {
    let items: [ProjectActivityEvent]
    let total: Int?
    let page: Int?
    let limit: Int?
}

struct ProjectMembership: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let spaceId: String
    let agentId: String?
    let userId: String?
    let role: String
    let isActive: Bool
    let roleLabel: String?
    let responsibility: String?
    let personaOverride: String?
    let isPrimary: Bool?
    let joinedAt: String?
    let updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case id, role
        case spaceId = "space_id"
        case agentId = "agent_id"
        case userId = "user_id"
        case isActive = "is_active"
        case roleLabel = "role_label"
        case responsibility
        case personaOverride = "persona_override"
        case isPrimary = "is_primary"
        case joinedAt = "joined_at"
        case updatedAt = "updated_at"
    }
}

struct ProjectMembershipListResponse: Decodable, Sendable {
    let memberships: [ProjectMembership]
    let total: Int?
}

struct ProjectParticipant: Identifiable, Hashable, Sendable {
    enum Kind: String, Hashable, Sendable {
        case member
        case agent
    }

    let id: String
    let name: String
    let kind: Kind
    let role: String
    let roleLabel: String?
    let responsibility: String?
    /// 人类成员用户 id；Agent 为 nil。点成员行开私信用。
    let userId: String?
    let agentId: String?
    let ownedByCurrentUser: Bool
    let isPrimary: Bool
}

File: apps/tabtin-android/app/src/main/java/com/tabtin/mobile/data/model/ProjectModels.kt
package com.tabtin.mobile.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject

/**
 * Project 是云端团队协作场；后端物理实现仍是 Space(type=team_space)。
 * Project 自身不拥有本地目录或移动端设备，执行发生在成员自己的 companion Workspace。
 */
@Serializable
public data class Project(
    val id: String,
    @SerialName("organization_id") val organizationId: String,
    val type: String? = null,
    val name: String,
    val description: String? = null,
    val avatar: String? = null,
    val color: String? = null,
    val status: String? = null,
    @SerialName("is_archived") val isArchived: Boolean? = null,
    val visibility: String? = null,
    @SerialName("member_count") val memberCount: Int? = null,
    @SerialName("primary_agent_id") val primaryAgentId: String? = null,
    @SerialName("can_manage") val canManage: Boolean = false,
    @SerialName("last_activity_at") val lastActivityAt: String? = null,
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("updated_at") val updatedAt: String? = null,
    @SerialName("my_workspace") val myWorkspace: ProjectCompanionWorkspace? = null,
) {
    public val displayDescription: String?
        get() = description?.trim()?.takeIf { it.isNotEmpty() }
}

@Serializable
public data class ProjectCompanionWorkspace(
    val id: String,
    val name: String? = null,
    @SerialName("agent_id") val agentId: String? = null,
    @SerialName("execution_agent_id") val executionAgentId: String? = null,
    @SerialName("working_dir") val workingDir: String? = null,
    @SerialName("control_device_id") val controlDeviceId: String? = null,
    @SerialName("control_device_status") val controlDeviceStatus: String? = null,
    @SerialName("is_companion") val isCompanion: Boolean? = null,
)

@Serializable
public data class ProjectListResponse(
    val projects: List<Project> = emptyList(),
    val total: Int = 0,
)

@Serializable
public data class PendingProjectInvitation(
    @SerialName("project_id") val projectId: String,
    @SerialName("project_name") val projectName: String,
    @SerialName("organization_id") val organizationId: String,
    val role: String,
    @SerialName("inviter_name") val inviterName: String,
    @SerialName("invited_at") val invitedAt: String? = null,
)

@Serializable
public data class PendingProjectInvitationListResponse(
    val invitations: List<PendingProjectInvitation> = emptyList(),
    val total: Int = 0,
)

@Serializable
public data class ProjectDiscussion(
    val id: String,
    @SerialName("organization_id") val organizationId: String,
    @SerialName("space_id") val spaceId: String? = null,
    @SerialName("space_name") val spaceName: String? = null,
    @SerialName("is_team_space_channel") val isTeamSpaceChannel: Boolean = false,
    val name: String = "",
    @SerialName("member_count") val memberCount: Int? = null,
    @SerialName("is_archived") val isArchived: Boolean = false,
    @SerialName("last_message_at") val lastMessageAt: String? = null,
    @SerialName("last_message_preview") val lastMessagePreview: String? = null,
    @SerialName("unread_count") val unreadCount: Int = 0,
)

@Serializable
public data class ProjectActivityEvent(
    val id: String,
    @SerialName("event_type") val eventType: String,
    @SerialName("actor_user_id") val actorUserId: String? = null,
    @SerialName("actor_name") val actorName: String? = null,
    @SerialName("target_type") val targetType: String? = null,
    @SerialName("target_id") val targetId: String? = null,
    @SerialName("target_name") val targetName: String? = null,
    val metadata: JsonObject? = null,
    @SerialName("created_at") val createdAt: String,
)

@Serializable
public data class ProjectActivityListResponse(
    val items: List<ProjectActivityEvent> = emptyList(),
    val total: Int = 0,
    val page: Int = 1,
    val limit: Int = 50,
)

@Serializable
public data class ProjectMembership(
    val id: String,
    @SerialName("space_id") val spaceId: String,
    @SerialName("agent_id") val agentId: String? = null,
    @SerialName("user_id") val userId: String? = null,
    val role: String,
    @SerialName("is_active") val isActive: Boolean = true,
    @SerialName("role_label") val roleLabel: String? = null,
    val responsibility: String? = null,
    @SerialName("persona_override") val personaOverride: String? = null,
    @SerialName("is_primary") val isPrimary: Boolean = false,
    @SerialName("joined_at") val joinedAt: String? = null,
    @SerialName("updated_at") val updatedAt: String? = null,
)

@Serializable
public data class ProjectMembershipListResponse(
    val memberships: List<ProjectMembership> = emptyList(),
    val total: Int = 0,
)

public enum class ProjectParticipantKind { MEMBER, AGENT }

public data class ProjectParticipant(
    val id: String,
    val name: String,
    val kind: ProjectParticipantKind,
    val role: String,
    val roleLabel: String? = null,
    val responsibility: String? = null,
    /** 人类成员用户 id；Agent 行为 null。点成员行开私信用。 */
    val userId: String? = null,
    val agentId: String? = null,
    val ownedByCurrentUser: Boolean = false,
    val isPrimary: Boolean = false,
)

@Serializable
public data class SetProjectPrimaryAgentRequest(
    @SerialName("agent_id") val agentId: String?,
)

public data class ProjectDetailData(
    val project: Project,
    val discussions: List<ProjectDiscussion> = emptyList(),
    val assets: List<SpaceResource> = emptyList(),
    val activities: List<ProjectActivityEvent> = emptyList(),
    val participants: List<ProjectParticipant> = emptyList(),
    val hasPartialFailure: Boolean = false,
    /**
     * 详情（getProject，my_workspace 的唯一来源）本身是否加载失败。
     * 与 hasPartialFailure 区分：其它分段失败不应把「执行环境」误判为「未准备」。
     */
    val detailFailed: Boolean = false,
)
